### ISSUE TITLE
POST file upload crashes on cgi.FieldStorage() [Python 3.4]

### ISSUE THREAD

**Issue by stephanepechard** [bottlepy/bottle#856]
With an [Android app](http://photobackup.github.io) using [OkHttp](http://square.github.io/okhttp/) for file upload through a POST request, I get the following error:

```
Traceback (most recent call last):
  File "/home/stephane/code/photobackup/server-bottle/venv/bin/bottle.py", line 862, in _handle
    return route.call(**args)
  File "/home/stephane/code/photobackup/server-bottle/venv/bin/bottle.py", line 1732, in wrapper
    rv = callback(*a, **ka)
  File "/home/stephane/code/photobackup/server-bottle/venv/lib/python3.4/site-packages/photobackup_bottle/photobackup.py", line 102, in save_image
    validate_password(request)
  File "/home/stephane/code/photobackup/server-bottle/venv/lib/python3.4/site-packages/photobackup_bottle/photobackup.py", line 78, in validate_password
    password = request.forms.get('password').encode('utf-8')
  File "/home/stephane/code/photobackup/server-bottle/venv/bin/bottle.py", line 166, in __get__
    if key not in storage: storage[key] = self.getter(obj)
  File "/home/stephane/code/photobackup/server-bottle/venv/bin/bottle.py", line 1084, in forms
    for name, item in self.POST.allitems():
  File "/home/stephane/code/photobackup/server-bottle/venv/bin/bottle.py", line 166, in __get__
    if key not in storage: storage[key] = self.getter(obj)
  File "/home/stephane/code/photobackup/server-bottle/venv/bin/bottle.py", line 1232, in POST
    data = cgi.FieldStorage(**args)
  File "/usr/lib/python3.4/cgi.py", line 559, in __init__
    self.read_multi(environ, keep_blank_values, strict_parsing)
  File "/usr/lib/python3.4/cgi.py", line 714, in read_multi
    self.encoding, self.errors)
  File "/usr/lib/python3.4/cgi.py", line 561, in __init__
    self.read_single()
  File "/usr/lib/python3.4/cgi.py", line 724, in read_single
    self.read_binary()
  File "/usr/lib/python3.4/cgi.py", line 746, in read_binary
    self.file.write(data)
TypeError: must be str, not bytes
192.168.0.21 - - [06/Jun/2016 20:58:04] "POST / HTTP/1.1" 500 744
```

It seems related to #567 (something around cgi.FieldStorage) but I don't see what I can do to fix the problem. Any idea?

There are two parameters and a file in the request. An example of headers is:

- User-Agent: PhotoBackup Android Client v0.8.1-dev
- Via: 1.1 vegur
- Host: requestb.in
- Accept-Encoding: gzip
- Content-Length: 1795216
- Total-Route-Time: 0
- X-Request-Id: 08838f63-da70-4318-8c78-665fb8ddcd99
- Content-Type: multipart/form-data; boundary=7f03fd45-e2b9-4f68-9d7d-50fd64dcb0a1
- Connect-Time: 1
- Connection: close

**Comment by defnull** [comment]
Looks like a bug in cgi.FieldStorage (again).

If a multipart section has a "Content-Length" header, but no "filename" attribute in the "Content-Disposition" header, cgi.FieldStorage tries to write binary data to a buffer file opened in text-mode.

The problem here is that cgi.FieldStorage has to decide if something is a binary file-upload or a unicode form-field, but it does so based on two different headers in two different places. If the headers contradict each other (form-fields usually don't have a Content-Length, file-uploads usually have a filename), it breaks.

Here is what happens:

A "Content-Length" header causes `cgi.FieldStorage.length` to be is set (which is fine).
https://hg.python.org/cpython/file/3.4/Lib/cgi.py#l550

If `length` has a value, `read_binary()` is used instead of `read_lines()` (which is questionable).
https://hg.python.org/cpython/file/3.4/Lib/cgi.py#l733

`read_binary()` calls `make_file()` which creates the buffer file in text mode if it does not find a `filename` attribute in the "Content-Disposition" Header (which is somewhat okay).
https://hg.python.org/cpython/file/3.4/Lib/cgi.py#l515
https://hg.python.org/cpython/file/3.4/Lib/cgi.py#l893

The bug is triggered if the last two steps disagree on the bytes vs. text question.


Sooo... for bottle. We could write another workaround. But actually I'd like to see this fixed upstream, really.

**Comment by defnull** [comment]
http://bugs.python.org/issue27308

**Comment by stephanepechard** [comment]
@defnull Thanks on your work on this bug, I didn't have the time (and probably neither the skills) to go that far.
I will follow the bug you opened on Python bug tracker.
As a workaround for my own use, I will maybe try to see if I can play with headers in my requests.

**Comment by defnull** [comment]
Looks like you just have to update your python installation. It appears to be fixed in 3.4, 3.5 and up. See http://bugs.python.org/issue24764#msg248778

The bug hunt was still fun :)

**Comment by stephanepechard** [comment]
Thanks for this!
I'm a bit disappointed I didn't find it myself while I was looking for it though...
I'll have to update my local Python 3.4 version (currently 3.4.3 on an old 14.10 Ubuntu)
